# STA 141B: Homework 1
Winter 2018

## Information

After the colons (in the same line) please write just your first name, last name, and the 9 digit student ID number below.

First Name: Wei-Kuang

Last Name: Lin 

Student ID: 997062562

## Instructions

We use a script that extracts your answers by looking for cells in between the cells containing the exercise statements.  So you 

- MUST add cells in between the exercise statements and add answers within them and
- MUST NOT modify the existing cells, particularly not the problem statement

To make markdown, please switch the cell type to markdown (from code) - you can hit 'm' when you are in command mode - and use the markdown language.  For a brief tutorial see: https://daringfireball.net/projects/markdown/syntax


## Part 1: The Doomsday Algorithm

The Doomsday algorithm, devised by mathematician J. H. Conway, computes the day of the week any given date fell on. The algorithm is designed to be simple enough to memorize and use for mental calculation.

__Example.__ With the algorithm, we can compute that July 4, 1776 (the day the United States declared independence from Great Britain) was a Thursday.

The algorithm is based on the fact that for any year, several dates always fall on the same day of the week, called the <em style="color:#F00">doomsday</em> for the year. These dates include 4/4, 6/6, 8/8, 10/10, and 12/12.

__Example.__ The doomsday for 2016 is Monday, so in 2016 the dates above all fell on Mondays. The doomsday for 2017 is Tuesday, so in 2017 the dates above will all fall on Tuesdays.

The doomsday algorithm has three major steps:

1. Compute the anchor day for the target century.
2. Compute the doomsday for the target year based on the anchor day.
3. Determine the day of week for the target date by counting the number of days to the nearest doomsday.

Each step is explained in detail below.

### The Anchor Day

The doomsday for the first year in a century is called the <em style="color:#F00">anchor day</em> for that century. The anchor day is needed to compute the doomsday for any other year in that century. The anchor day for a century $c$ can be computed with the formula:
$$
a = \bigl( 5 (c \bmod 4) + 2 \bigr) \bmod 7
$$
The result $a$ corresponds to a day of the week, starting with $0$ for Sunday and ending with $6$ for Saturday.

__Note.__ The modulo operation $(x \bmod y)$ finds the remainder after dividing $x$ by $y$. For instance, $12 \bmod 3 = 0$ since the remainder after dividing $12$ by $3$ is $0$. Similarly, $11 \bmod 7 = 4$, since the remainder after dividing $11$ by $7$ is $4$.

__Example.__ Suppose the target year is 1954, so the century is $c = 19$. Plugging this into the formula gives
$$a = \bigl( 5 (19 \bmod 4) + 2 \bigr) \bmod 7 = \bigl( 5(3) + 2 \bigr) \bmod 7 = 3.$$
In other words, the anchor day for 1900-1999 is Wednesday, which is also the doomsday for 1900.

__Exercise 1.1.__ Write a function that accepts a year as input and computes the anchor day for that year's century. The modulo operator `%` and functions in the `math` module may be useful. Document your function with a docstring and test your function for a few different years.  Do this in a new cell below this one.

In [171]:
import math  # load package "math", so I can use math.floor fucntion
import math 
from datetime import datetime
import operator # find the index of max and min in a set

In [172]:
def anchor_day(year):
    """ Find the anchor day for the given year
    
    Arguemnts: year
    
    Returns: The doomsday for the given year. The output is the 
    integer number form 0 to 6.  For example 0 = Sunday, 1= Monday, ...
    """
    c= math.floor(year / 100) 
    a = (5*(c % 4) +2 ) % 7
    return a

##### list  of anchor days for each centries
#### 1800 - 1899: Friday
#### 1900 - 1999: Wednesday
#### 2000 - 2099: Tuesday
#### 2100 - 2199: Sunday

In [5]:
# 1.1 test
anchor_day(1801)

5

In [394]:
# 1.1 test
anchor_day(1999)

3

In [6]:
# 1.1 test
anchor_day(1901)

3

In [179]:
# 1.1 test
anchor_day(1981)

3

In [173]:
def centry(year):
    """ Find the first two digit of the given year
    
    Arguemnts: year
    
    Returns: the first two digit of the given year
    """
    c= math.floor(year / 100) 
    return c

In [395]:
# 1.1 test
centry(1998)

19

In [396]:
# 1.1 test
centry(2012)

20

### The Doomsday

Once the anchor day is known, let $y$ be the last two digits of the target year. Then the doomsday for the target year can be computed with the formula:
$$d = \left(y + \left\lfloor\frac{y}{4}\right\rfloor + a\right) \bmod 7$$
The result $d$ corresponds to a day of the week.

__Note.__ The floor operation $\lfloor x \rfloor$ rounds $x$ down to the nearest integer. For instance, $\lfloor 3.1 \rfloor = 3$ and $\lfloor 3.8 \rfloor = 3$.

__Example.__ Again suppose the target year is 1954. Then the anchor day is $a = 3$, and $y = 54$, so the formula gives
$$
d = \left(54 + \left\lfloor\frac{54}{4}\right\rfloor + 3\right) \bmod 7 = (54 + 13 + 3) \bmod 7 = 0.
$$
Thus the doomsday for 1954 is Sunday.

__Exercise 1.2.__ Write a function that accepts a year as input and computes the doomsday for that year. Your function may need to call the function you wrote in exercise 1.1. Make sure to document and test your function.

In [184]:
def doomsday(year):
    """ Find the doomsday for the given year
    
    Arguemnts: year
    
    Returns: The doomsday for the given year. The output is the 
    integer number form 0 to 6.  For example 0 = Sunday, 1= Monday, ..., 6= Saturday
    """
    c= math.floor(year / 100) # c is the centry
    a = (5*(c % 4) +2 ) % 7 # a is the anchor year
    y= year % 100 # y is the last two digit
    doomsday= (y+ math.floor(y/4)+a)% 7
    return doomsday

In [175]:
#  1.2 test
doomsday(1954) # it works!

0

In [80]:
# 1.2 test
doomsday(1998) 

6

In [81]:
# 1.2 test
doomsday(2001)

3

### The Day of Week

The final step in the Doomsday algorithm is to count the number of days between the target date and a nearby doomsday, modulo 7. This gives the day of the week.

Every month has at least one doomsday:
* (regular years) 1/10, 2/28
* (leap years) 1/11, 2/29
* 3/21, 4/4, 5/9, 6/6, 7/11, 8/8, 9/5, 10/10, 11/7, 12/12

__Example.__ Suppose we want to find the day of the week for 7/21/1954. The doomsday for 1954 is Sunday, and a nearby doomsday is 7/11. There are 10 days in July between 7/11 and 7/21. Since $10 \bmod 7 = 3$, the date 7/21/1954 falls 3 days after a Sunday, on a Wednesday.

__Exercise 1.3.__ Write a function to determine the day of the week for a given day, month, and year. Be careful of leap years! Your function should return a string such as "Thursday" rather than a number. As usual, document and test your code.

In [399]:
def day_fun(data_input):
    """ Find the weekday for the given date 'm/d/Y' (m=month, d= day, Y=year)
    
    Arguemnts: The input is the requested date. The formate is 'm/d/Y' ex: '02/14/2018' 
    
    Returns: The doomsday for the given year. The output is the 
    integer number form 0 to 6.  For example 0 = Sunday, 1= Monday, ...
    """

    data = data_input.split('/')
    
    month= int(data[0])
    day  = int(data[1])
    year = int(data[2])
    
    # PartA: To find the nearest Doomsday
    regular = [10, 28, 21, 4, 9, 6, 11, 8, 5, 10, 7, 12]
    leap= [11, 15, 21, 4, 9, 6, 11, 8, 5, 10, 7, 12] # Feb.15th is also the doomsday in leap year
    weekday= ["Sunday","Monday","Tuesday", "Wednesday", "Thursday", "Friday","Saturday"]
    
    # Convert the date into the the caculabe data, so I could find the nearest doomsday across month
    date_format = '%m/%d'
    regular_date= [datetime.strptime('1/10', date_format),
            datetime.strptime('2/28', date_format),
            datetime.strptime('3/21', date_format),
            datetime.strptime('4/4', date_format),
            datetime.strptime('5/9', date_format),
            datetime.strptime('6/6', date_format),
            datetime.strptime('7/11', date_format),
            datetime.strptime('8/8', date_format),
            datetime.strptime('9/5', date_format),
            datetime.strptime('10/10', date_format),
            datetime.strptime('11/7',date_format),
            datetime.strptime('12/12',date_format)]
    
    leap_date= [datetime.strptime('1/11', date_format),
            datetime.strptime('2/15', date_format),
            datetime.strptime('3/21', date_format),
            datetime.strptime('4/4', date_format),
            datetime.strptime('5/9', date_format),
            datetime.strptime('6/6', date_format),
            datetime.strptime('7/11', date_format),
            datetime.strptime('8/8', date_format),
            datetime.strptime('9/5', date_format),
            datetime.strptime('10/10', date_format),
            datetime.strptime('11/7',date_format),
            datetime.strptime('12/12',date_format)]
    
    # recreate m/d from input data
    md = (str(month) + '/' + str(day) )
    date_format = '%m/%d'
    md_num = datetime.strptime(md, date_format)
    
    #Determine the doomsday
    doom = doomsday(year)
    
    #Find the nearest doomsday, given non-leap or leap year
    if (year%4 + year%400) ==0:
        # print("it is a leap year")
        
        # find the nearest Doomsday
        # Calculate diff_doom (difference between md_num and each doomsday)
        diff_doom = [ abs(leap_date[x] -md_num) for x in range(12)]
        
        # Result of nearest Doomsday
        min_index, min_value = min(enumerate(diff_doom), key=operator.itemgetter(1))
        nearst_day= leap[min_index]
        
        date_format_mdY = '%m/%d/%Y'
        input_num = datetime.strptime(data_input, date_format_mdY) # convert it into caluable value
        nearst_day_A = str(min_index +1) + '/' + str(nearst_day) + '/' + str(year) # reassembel nearst doomsday in the given year
        nearst_day_num = datetime.strptime(nearst_day_A , date_format_mdY) # convert it into caluable value
        
        diff_num = input_num - nearst_day_num 
        diff = diff_num.days
        
        if diff <0:
            movement = (-1)*( (abs(diff))%7 )
            day_of_week= doom + movement #
            
            if day_of_week <0:
                day_of_week= (-1)* ( abs(day_of_week)%7 )
                result = weekday[day_of_week]
            else:
                result= weekday[day_of_week%7]
            
        else: # When diff is positive
            movement = diff %7 
            day_of_week= doom + movement
            result= weekday[day_of_week%7]
        
    else:
        # print("it is a regular year")
        # find the nearest Doomsday
        # Calculate diff_doom (difference between md_num and each doomsday)
        diff_doom = [ abs(regular_date[x] -md_num) for x in range(12)]
        
        
        # Result of nearest Doomsday
        min_index, min_value = min(enumerate(diff_doom), key=operator.itemgetter(1))
        nearst_day= regular[min_index]
        
        date_format_mdY = '%m/%d/%Y'
        input_num = datetime.strptime(data_input, date_format_mdY) # convert it into caluable value
        nearst_day_A = str(min_index +1) + '/' + str(nearst_day) + '/' + str(year) # reassembel nearst doomsday in the given year
        nearst_day_num = datetime.strptime(nearst_day_A , date_format_mdY) # convert it into caluable value
        
        diff_num = input_num - nearst_day_num 
        diff = diff_num.days
        
        if diff <0:
            movement = (-1)*( (abs(diff))%7 )
            day_of_week= doom + movement 
            
            if day_of_week <0:
                day_of_week= (-1)* ( abs(day_of_week)%7 )
                result = weekday[day_of_week]
            else:
                result= weekday[day_of_week%7]
        else: # When diff is positive
            movement = diff %7 
            day_of_week= doom + movement # 
            result= weekday[day_of_week%7]
            
    return result

In [400]:
# 1.3 test
day_fun('1/1/2001') # it's correct: Jan 1st , 2001 is Monday

'Monday'

In [401]:
# 1.3 test
day_fun('9/29/1998') # it's correct: Sep 29 , 1998 is Tuesday

'Tuesday'

In [402]:
# 1.3 test
day_fun('9/29/2000') # it's correct: Sep 29 , 2000 is Friday

'Friday'

In [393]:
# 1.3 test
day_fun('1/13/1900') # it's correct: Jan 13 , 1900 is Saturday

'Saturday'

__Exercise 1.4.__ How many times did Friday the 13th occur in the years 1900-1999? Does this number seem to be similar to other centuries?

In [200]:
def Friday_count_1(c1,c2):
    """ Count the number of 13th Friday in the requested interval 
    
    Arguemnts: The input is the requested interval, in terms of year. 
    c1 is the lower end of the interval 
    c2 is the upper end of the interval
    type of c1 and c2 are numeric 
    
    Returns: Sum of the number of 13th Friday in the requested interval 
    """
    
    i =0
    for year in range(c1,c2+1):
        for month in range (1,13):
            data= str(month) + '/'+ '13' + '/' + str(year)
            if day_fun(data) == 'Friday':
                i = i+1
            else:
                i=i
            
    return i

In [205]:
# The number of 13th Friday in centries from 18th, 19th, 20th, 21th
Friday_count_1(1700,1799), Friday_count_1(1800,1899), Friday_count_1(1900,1999), Friday_count_1(2000,2099)

(171, 172, 173, 170)

1.4 ANS:
The number of 13th Friday seems to be similar among different centries
18th  19th  20th  21th
(171,  172,  173,  170) 

__Exercise 1.5.__ How many times did Friday the 13th occur between the year 2000 and today?

In [333]:
def Friday_count_2( sdy, today):
    """ Count the number of "13th Friday" between a assigned year and today
    
    Arguments: 
    sdy is the start year, format= 'Y', ex: '2000'  
    today is the date of toda, format = '%m/%d/%Y', ex: '1/19/2018'
        
    Returns:
        the number of "13th Friday" within the assigned period
    """""        
    tdy = int( today.split('/')[2] )  # tdy = the year of today
    tdm = int( today.split('/')[0] )  # tdm = the month of today
    tdd = int( today.split('/')[0] )  # tdd = the day of today
    sdy = int(sdy)
    
    #Part-a  includes the date from year_start to the previous year
    a = 0
    for Y in range(sdy, tdy):
        for M in range (1,13):
            data= str(M) + '/'+ '13' + '/' + str(Y)
            if day_fun(data) == 'Friday':
                a = a+1
            else:
                a=a
            
    #Part-b  includes dates in this year unti previous month
    b = 0
    for M in range(1,tdm):
        data= str(M) + '/'+ '13' + '/' + str(tdy)
        if day_fun(data) == 'Friday':  #'1/13/2018' = today
            b = b+1
        else:
            b=b
            
    #Part-c  includes dates in the month of today
    c = 0
    if (tdd >= 13 and day_fun(data) == 'Friday' ):
        c=1
        
            
    return a+b+c

In [334]:
Friday_count_2('2000','1/19/2018')

29

1.5 ANS:
Between 2000 and today, there were "29" 13th Friday.

## Part 2: 1978 Birthdays

__Exercise 2.1.__ The file `birthdays.txt` contains the number of births in the United States for each day in 1978. Inspect the file to determine the format. Note that columns are separated by the tab character, which can be entered in Python as `\t`. Write a function that uses iterators and list comprehensions with the string methods `split()` and `strip()` to  convert each line of data to the list format

```Python
[month, day, year, count]
```
The elements of this list should be integers, not strings. The function `read_birthdays` provided below will help you load the file.

In [335]:
def read_birthdays(file_path):
    """Read the contents of the birthdays file into a string.
    
    Arguments:
        file_path (string): The path to the birthdays file.
        
    Returns:
        string: The contents of the birthdays file.
    """
    with open(file_path) as file:
        return file.read()

In [369]:
def converter(file_path):
    """ Convert the contents of the birthdays file into one huge list. 
        In this list, each of element includes [month, day, year, count]
    
    Arguments:
        file_path (string): The path to the birthdays file.
        
    Returns:
        A list, each of its element includes [month, day, year, count]
    """
    
    data = read_birthdays (file_path).strip()
    result = [0]* 365
    i= 0
    
    for line in data.split('\n')[6:]:
        date, count = line.split('\t')
        month, day, year = date.split('/')
        ListA = [int(month), int(day), int(year), int(count)]
        result[i] =  ListA 
        i = i +1
    return(result)

In [372]:
file_path = "C:/Users/WeiKuang/Documents/GitHub/141b-hw1-Wei-Kuang/birthdays.txt"
result_21 = converter(file_path)

# To print the first 9 data in the file
for i in range(9):
    print(result_21[i])

[1, 1, 78, 7701]
[1, 2, 78, 7527]
[1, 3, 78, 8825]
[1, 4, 78, 8859]
[1, 5, 78, 9043]
[1, 6, 78, 9208]
[1, 7, 78, 8084]
[1, 8, 78, 7611]
[1, 9, 78, 9172]


__Exercise 2.2.__ Which month had the most births in 1978? Which day of the week had the most births? Which day of the week had the fewest? What conclusions can you draw? You may find the `Counter` class in the `collections` module useful.

In [373]:
# '1/1/1978' is Sunday. Thus, the sequence of the weekday can starts from "Sunday"
# Set Sunday = 0, Monday =1, and so forth
weekday= [w for w in range(0,7)]

# weekday*52 has the first 364 weekdays. There is one weekday left, which is Sunday that represents 0
All_weekday = weekday*52 + [0]

# Insert weekday info to the fist poistion of each element in the "result_21" list
for x in range(0,365):
    result_21[x].insert(0, All_weekday[x])

# print the expaned result
for x in range(14):
    print(result_21[x]) 
    # [weekday, month, day, count]

[0, 1, 1, 78, 7701]
[1, 1, 2, 78, 7527]
[2, 1, 3, 78, 8825]
[3, 1, 4, 78, 8859]
[4, 1, 5, 78, 9043]
[5, 1, 6, 78, 9208]
[6, 1, 7, 78, 8084]
[0, 1, 8, 78, 7611]
[1, 1, 9, 78, 9172]
[2, 1, 10, 78, 9089]
[3, 1, 11, 78, 9210]
[4, 1, 12, 78, 9259]
[5, 1, 13, 78, 9138]
[6, 1, 14, 78, 8299]


In [382]:
# sum by month
SM= [0]*12

for item in result_21:
    for m in range(1,13):  # Be careful, m is going to compare with the Jan(1), Feb(2), March(3),..
        if item[1] == m:
            SM[m-1] = SM[m-1] + item[4]
            
SM # It has the sum of births by month. The order is [Jan. Feb. Mar. ...]

[270695,
 249875,
 276584,
 254577,
 270812,
 270756,
 294701,
 302795,
 293891,
 288955,
 274671,
 284927]

In [386]:
### ANS to 2.2 - 1
max_index, max_value = max(enumerate(SM), key=operator.itemgetter(1))

max_index +1 , max_value # Month (8 = Agust) with most births, the max value

(8, 302795)

In [377]:
# sum by week
total = [0,0,0,0,0,0,0] 

for item in result_21:
    for w in range(0,7):
        if item[0] == w:
            total[w] = total[w] + item[4]
    

total # It has the total Birthday counts for each weekday.  [Sun, Mon, Tu, ...]

[421400, 487309, 504858, 493897, 493149, 500541, 432085]

In [ ]:
### ANS to 2.2 - 2
# Sunday has least births.
# Tuesday has the most births.

__Exercise 2.3.__ What would be an effective way to present the information in exercise 2.2? You don't need to write any code for this exercise, just discuss what you would do.

Ans 2.3:

1. Showing a table with the proportion of the births by groups (ex: Weekday-Sun, Mon,... or Month-Jan, Feb, ...) can be good. However, if we want to show the births by weekday and month at once, a mosaic plot with two factors-levels is definetly an effective way to spot the lowest and highest porprtion of birth under certain weekday-month combinations.

2. Another way to visualise a moasic plot could be heat map, the density of the color reflect the porportion of biths in all weekday-month combinations.

3. Showing a bar chart with well-explanied lables could be another good way to show the quantity in different Weekdays or Months.

